# Project title    

### Author: Muhammad kala     
### Ecological Genomics:   

## Overall Description of notebook      

Notebook for the Ecological Genomics Class. I still prefer my physical notebook :(


## Date started: (2017-01-18)   
## Date end:   (2017-month-day)    

## Philosophy   
Science should be reproducible and one of the best ways to achieve this is by logging research activities in a notebook. Because science/biology has increasingly become computational, it is easier to document computational projects in an electronic form, which can be shared online through Github.    

### Helpful features of the notebook     

**It is absolutely critical for your future self and others to follow your work.**     

* The notebook is set up with a series of internal links from the table of contents.    
* All notebooks should have a table of contents which has the "Entry", date, and title (information that allows the reader to understand your work).     
* Also, one of the perks of keeping all activities in a single document is that you can **search and find elements quickly**.     
* You can document anything you'd like, aside from logging your research activities. For example:
	* feel free to log all/any ideas for your research project([example](https://github.com/adnguyen/Notebooks_and_Protocols/blob/master/2016_notebook.md#Entry-39-2016-06-13-post-doc-project-idea-assessing-current-impacts-of-climate-change-in-natural-populations)) as an entry,     
	* or write down notes for a paper([example](https://github.com/adnguyen/Notebooks_and_Protocols/blob/master/2016_notebook.md#id-section36).      

* Lastly, you can share specific entries because of the three "#" automatically creates a link when the notebook renders on github.      


<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.  


### Table of contents for 60 entries (Format is *Entry: Date(with year-month-day). Title*)        
* [Entry 1: 2017-01-18](#id-section1). First class; intros
* [Entry 2: 2017-01-23](#id-section2). 
* [Entry 3: 2017-01-25](#id-section3).
* [Entry 4: 2017-01-30](#id-section4).
* [Entry 5: 2017-02-01](#id-section5).
* [Entry 6: 2017-02-06](#id-section6).
* [Entry 7: 2017-02-08](#id-section7).
* [Entry 8: 2017-02-13](#id-section8).
* [Entry 9: 2017-02-15 and 2017-02-22](#id-section9).
* [Entry 10:](#id-section10).
* [Entry 11:](#id-section11).
* [Entry 12:](#id-section12).
* [Entry 13:](#id-section13).
* [Entry 14:](#id-section14).
* [Entry 15:](#id-section15).
* [Entry 16:](#id-section16).
* [Entry 17:](#id-section17).
* [Entry 18:](#id-section18).
* [Entry 19:](#id-section19).
* [Entry 20:](#id-section20).
* [Entry 21:](#id-section21).
* [Entry 22:](#id-section22).
* [Entry 23:](#id-section23).
* [Entry 24:](#id-section24).
* [Entry 25:](#id-section25).
* [Entry 26:](#id-section26).
* [Entry 27:](#id-section27).
* [Entry 28:](#id-section28).
* [Entry 29: Code for HW2](#id-section29).
* [Entry 30: Code for HW3](#id-section30).
* [Entry 31: Scripts for Project (microbiome)](#id-section31).
* [Entry 32:](#id-section32).
* [Entry 33:](#id-section33).
* [Entry 34:](#id-section34).
* [Entry 35:](#id-section35).
* [Entry 36:](#id-section36).
* [Entry 37:](#id-section37).
* [Entry 38:](#id-section38).
* [Entry 39:](#id-section39).
* [Entry 40:](#id-section40).
* [Entry 41:](#id-section41).
* [Entry 42:](#id-section42).
* [Entry 43:](#id-section43).
* [Entry 44:](#id-section44).
* [Entry 45:](#id-section45).
* [Entry 46:](#id-section46).
* [Entry 47:](#id-section47).
* [Entry 48:](#id-section48).
* [Entry 49:](#id-section49).
* [Entry 50:](#id-section50).
* [Entry 51:](#id-section51).
* [Entry 52:](#id-section52).
* [Entry 53:](#id-section53).
* [Entry 54:](#id-section54).
* [Entry 55:](#id-section55).
* [Entry 56:](#id-section56).
* [Entry 57:](#id-section57).
* [Entry 58:](#id-section58).
* [Entry 59:](#id-section59).
* [Entry 60:](#id-section60).

------
<div id='id-section1'/>
### Entry 1: 2016-07-18. Ecological genomics, first class (Andrew's notes)

### **Steve and Melissa's intro**    
* Steve: It is a young field, trying to establish it's own identity    
	* Ecological genomics institute, KSU: emphasis on adaptation to environment   
	* Gordon Research Conference: Integrating different levels of biological organization on **ANY SYSTEM**; approach and tool focused! Field going towards new data and new analytic techniques  
	* Intro to eco genomics, oxford press; Using technology to address ecological issues such as nutrient cycling, population structure, life history vairation , trophic interaction, stress responess, and adpatation to environmental change   

* DATA driven: next gen sequencing revolutionizes biology
	* creats a new problem--large datasets!!! how to make sense? 
	* not data limited and potentially computationally limited   

* Where is the field headed    
	* Molecular Ecology Journal(flagship journal representative o the field)  
		* ALL systems:  corals, protists, daphnia, coral, lemurs, dandelions, steve studies trees 
		* model organism constraint disappearing!   
	* What types of questions are asked?  
		* How do genes correspond with circadian rythm?  
		* How does the microbiome influence the organism? 
		* How does epigenetic variation influence evolutionary responses? or contribute to phenotypic variation?  
		* What are the patterns of genetic diversity that can give us insights on population dynamics?  
		* What are constraints and tradeoffs and genetic mechanisms of traits? 

* Methods?   
	* De novo genome assembly; sequencing a DNA book from scratch!!    
		* RNA-seq; transcriptomic profiling     
	* 16 s metagenomic sequencing      
	* Rad-seq/GBS for estiamting population structure and genetic diversity     

* Proccesses studied?    
	* All evo and eco stuff; speciation, hybridization, local adaptation, genetic basis of local adaptation, genetic architecture of complex phenotypes, genes controlling host-pathogen evolutionary dynamics, pop structure, gene flow, epigenetics     
	
* Goals of the course!    
	1. Learn how ecology and genomes shape each other   
	2. Think creatively about major questions, and pose testable hypotheses to those questions using appropriate genomic data    
	3. Think about careful experimental design and statistical analysis---shown by reading papers   
	4. Achive working knowledge and level of comfort for bioinformatics routines for ecological genomics studies   
	
### Melissa background	  

**Background, what drove Melissa and Steve to ecological genomics?**       

Melissa read a cool paper that scales from analyzing a few loci to the whole genome.   

One figure popped out at her, FST (developed by Sewell Wright) histogram.   FST of 1= complete differentiation, FST of 0 = no diff. FST described as **Alleles in space**. From this histogram, Melissa was struck by how you can separate out neutral from selective ones.  

Melissa has a data set with 96 sea stars and then the 16s microbiome. Would be cool to see if there is heritability in some bactera

### Steve background   

* Inspired by Yanis Antonivics (an **OG**)   
* At the time, just so stories: **Adaptatationist programme**    
	* Just go out and go by feeling in a natural history way and prescribe an adaptation story   
	* Janis wrote a creed to quantify the operational relationship between traits, environment, and genetics     
* Yanis was on Steve's committee and Steve was interested in adaptation with respect to invasion biology because organisms need to respond to novel environments     
	* Phenotypes can relate to the environment, but what is the genetic basis of local adaptation (in situ)? There are other confounding issues: demographic effects, plasticity     
* Steve thinks about environment-phenotype-genetics triangle. Basically a path diagram that feeds back on each other.    
	* Relationship between genes and phenotype ---GWAS (Genome wide association study)    
	* Relationship between genetics and environment --- Fst, clines between allele frequencies and environment    
* Invasion history is tough because of demographic history    
* He decided to focus on trees; large population size, straddle huge environmental gradients so the opportunity for selection is high   
	* positive relationship between Growing season length and traits    
	* Did a  reciprocal transplant of different populations to identify the extent of local adaptation in large established common gardens    
	* SK does GBS (genotype by sequencing)      
	* Problem with field: validating key gene candidates            
  

------
<div id='id-section2'/>
### Entry 2: 2017-01-23

## Rubric for Info Update:
* Outline   
* ~20 min   
* Learning/engaging activity   
* Use board efficiently   
* Take-home messages   
* Examples from literature   

### Melissa's Info Update: Next Generation Sequencing

## 1. Advances in Technology  
Huge advances made in technology in the last decade (still moving quite fast)

Contrast between Human Genome Project and current technology

Human Genome Project:
* done 2001-2003   
* ABI-Sanger Sequencing   
* 15 years   
* 1 genome   
* $3 billion   

HiSeq XTen (Illumina):
* 2014   
* 1 day   
* 45 whole human genomes   
* S1000 each   

## 2. Range of Applications
* WGS (Whole Genome Sequencing)   
* RNA-seq   
* ChIP-seq (Chromatin Immunoprecipitation sequencing)   
* Targetted/Capture sequencing   

## 3. General Library Prep. Workflow   

#### Techniques depends on applications/goals

where is genetic variation?   

number of samples:
* population   
* individual   
* comparative studies   
* model organism or not (model organisms have reference genomes/transcriptomes etc)   

demographic history

adaptive genentic variation

gene expression variation

#### Decisions based on what is needed:

* Length of reads   
* Number of reads   
* Distribution of reads   

Reduced rep.:
* RNA (coding)-more reads for highly expressed regions   
* GBS/RAD (near restriction sites)    

Extraction (DNA or RNA->cDNA), fragment sample, ligate adaptors (individual barcodes), add sequence adaptors, PCR    

## 4. Sequence by Synthesis (SBS)

Adapters are added to our DNA segments. DNA goes into a lane of flow cell. Adapters bind to oligo's and our DNA forms a bridge. A picture is taken every time a complementary labeled dNTP's is added.   

90% of global data by Illumina     

Combining different methods help in assembling genomes of non-model organisms (de novo assembly)     

## Glossary 
```
short reads 50 bp   
long reads 100 bp, 150 bp, 300 bp    
longer reads 10000-60000 bp (SMRT)- faster but not as accurate, so not good for de novo assembly   
```

------
<div id='id-section3'/>
### Entry 3: 2017-01-25

## Info Update: QTN

## 1. What are QTN's?
Quantitative Trait Nucleotides
E.g. of Quantitative Traits   
-Flowering time (phenology)-continuous trait Â       
-Flower color-discrete trait       
-Thermal tolerance       
-Venom potency   
-Defense compounds   
-Toxin tolerance   
-Drought tolerance        
-Altitude (hypoxia) tolerance   

## 2. Quantitative genetic theory of adaptive traits   
V~A~ (additive genetic variance)   
h^2^ (heritability)   

Reminder h^2^=V~A~/V~P~, where V~P~ is phenotypic variance                 
Since mutation can diminish or augment avg fitness, big change mutations tend to be extreme and not selected. More info with figures in notebook

## 3. Methods
-Linkage mapping      
-Genome-Wide Association Studies (GWAS)     
-Selection Scans     




------
<div id='id-section4'/>
### Entry 4: 2017-01-30

## Project Info

## Sea Stars Wasting Disease

- Widespread   
- Many species affected (resistance vary among species)   
- Starts with loss of turgor pressure and/or lesions (holes)     
- Progress into limbs being pulled off (when sea stars move or through gravity)      
- *Pisaster* *ochraceus* (Alaska to Baja, CA). Orange or purple     
- First reports of disease in 2012 (Isolated cases from 1970)     
- Densovirus implicated     
- Present in 70 yo museum specimens      
- Outbreaks in 1970's, 1980's but not as extreme as current outbreak     
- All individuals exposed     
- Disease seem to be more prevalent intertidally (warmer temperatures, more variable environment)   
- Epidermal (host biopsy). Extracted RNA, got all mRNA (poly A-tail) using 5 Illumina hi-Seq 3000PE100 lanes to get gene expression data.   
- 16S rRNA amplified     
- Starfish collected and delivered within 24 hours of collection (all healthy at time)     
- Transmission might be environmental or through the gut     


## Update
Densovirus seem to be missing from samples

## Project ideas for each group in notebook and on [Andrew's github](https://github.com/adnguyen/Notebooks_and_Protocols/blob/master/2017_Eco_Gen_ANBE_nb.md#id-section4) 

Kirsten and I have chosen to look at SNPs to see if allelic differences might be accounting for resistance to disease, and maybe look at microbiome as well.

------
<div id='id-section5'/>
### Entry 5: 2017-02-01

## Info update: WGS    

### 1. Applications:    
-Population genetics: effective population size, demographic events (admixture), genetic relatedness, inbreeding    
-Conservation (genomics): monitoring, breeding programs    
-Taxonomic delineations    

Benefits 
* Higher power + resolution   
* New possible research areas: screening for variation, adaptive potential of population e.g.    
-QTL/Association mapping     
-pinpoint loci of adaptations      
-impact of genetic variation    
-patterns of gene expression    
-plastic response to environmental changes    
-use of ancient material    
-environmental samples 

### 2. Planning stage

Basic considerations 
- Is sequence necessary for purpose?    
- Do we have the resources for quality genome production?     
low quality = low coverage of whole genome    

Limitations (Usually one individual sequenced):
- Polymorphic genes    
- Paralogs    
- Rapidly evolving genes     
- Large gene families     
- Not type specimens (Use of pooled samples can help fix this)    
- Impossible to sequence all nucleotides (centromeres an telomeres; highly repetitive regions)     
- Error     
- Working hypothesis: assembly tools     

Prior Considerations     
- seq platform:     
-Short: Illumina Hi-seq (150bp) and SOLID (50bp)       
-Longer: Pacific Biosciences (5Kb), Ion torrent (~500bp), Illumina Molecule (up to 10KB)     
-Hybrids (studies typically use more than one i.e. both short and long reads for scaffolding)      

- Knowledge on organism     
-Genome size     
-Seq error rates    
-Repeat content     
-Degree of genome duplications     
Recommended pilot short-insert seq      
 
### 3. Methods

Wet lab procedures    
* Individual     
-Heterozygous position have adverse effects on assembly     
-Highly polypoid are challenging (special pipelines); use of inbred individuals, parthenogenic or genocentric offspring         
-Have metadata, age, sex, location   

* Tissue     
-Avoid energetically active tissue such as muslce (increase in mitochondrial DNA because increase in read depth)   
-Avoid gut+skin (non target DNA)        

* DNA      
-High quality, non degradable     
-Check with high resolution gel      

* Library Preparation     
I. Use single reads/remove duplicated reads    
II. Insert sizes within 0.2-40 kb range (>20 kb for scaffold)    
III. Read orientation:    
-Inwards (paired-end)      
-Outwards (outward)       
Mis orientation can happen/ could be chimeric segment from non adjascent DNA     

### 4. Genome assembly  

Storage     
Server     
Expertise    

Pipeline applicable to non models    

* Pre Assembly    
-Quality    
-GC content    
-Repeat abundance    
-Proportion of duplicate read     
-Seq for calibration must be removed (Illumina)    

* De novo     
  
-Algorithms (minimize misassembly and improve contiguity)    
-OLC (Overlap-Layout-Consensus)    
-Extension-based     
-Hybrid approaches    

## Info update: RNA-seq   

### 1. Advantages:
- Differential gene expression: in tissues there is variation    
- Allele specific expression: environmental response of adaptation       
- Functional relevant subset of the genome       


RNA seq better than microarray in wide range expression, info on splicing events     

### 2. Limitations

- Does not give abundance of protein or other post-translational modification of proteins (and function)    

### 3. Workflow

Exp setup -> Tissue prep + library -> High throughput seq -> Transcriptome reconstruction -> Alignment of reads -> Read Qualification/Marker development -> Biological Influence     

- Setup   
-Coding or regulatory      
-Is there a reference genome?     
-Alternative splicing?     
-Technology?     
-Population or specific treatments     

- Tissue prep     
Small organisms-many tissues pulled    
-Sample should be in RNase free environment    
-DNase treatment    
-get rid of rRNAs      
Figures in physical textbook.      

- Library    
-Single end     
-Paired end      

- High throughput seq     
Choose platform:      
-Pyrosequencing by Roche (some incorrect polymer)     
-Ion torrent      
-GA/Hi-seq Illumina     

- Reconstruction    
Figures in physical textbook.      

## Info Update: Amplicon seq     

Targetted approach for analyzing genetic variation in specific genomic region      

Amplicon is a targeted gene (region) to be amplified via PCR w/ specific primers      

- Library Prep     

Extract -> 1st PCR (amplify gene w/ primers) -> gel -> 2nd PCR w/ barcodes and adaptors (or directly to Seq. Facility) -> clean -> pool -> Seq facility     

454 replaced by Mi-Seq platform from Illumina  

Trim adaptors      
Align    

- Limitation   
Time consuming to manually correct some sequencing mistakes    
 
- Advantages    
Save time and money     

## Info Update: GBS RAD-seq

GBS: Genotyping by sequencing     
RAD: Restriction assisted DNA sequencing (using restriction enzymes)     

GBS RAD-seq is between RNA-seq and Amplicon in terms of completeness of data and sampling     

-lot of individuals     
-lots of SNPs across genomes     
-don't care about specific genes     
Useful for population genetics (population genetic diversity of one species)     

- Library Prep    

Restriction enzymes -> Barcodes + Adaptors -> PCR -> sequence    

More info and drawings in notebook.      


## Coding

Doing a de novo assembly on Trinity 

First, we will be cleaning files using Trimmomatic PE

Logging in to terminal (through PuTTY)

server pbio381.uvm.edu
Use UVM login and password as login and password

"top" command that tells us who is using cpus in the server

"q" quit

"ll" List all files in the current directory

"mkdir abc" makes a directory (folder) called "abc" where you are

"~" home directory (we all have one)

"cd" makes you change directory (so cd ~ makes you go to home directory)

"pw" tells us where we are (path from home directory)

"rm" remove file

"mv" move files    



[Unix Cheatsheet](https://files.fosswire.com/2007/08/fwunixref.pdf)



------
<div id='id-section6'/>
### Entry 6: 2017-02-06

## Info Update: RNA seq 2

Approach    
Experimental Design     
Library Prep     
Sequencing     
Receive Data    
Computer/ Server Setup     

- Workflow    
1. Clean reads (files: .fastq, . fq, .gz)      
-adaptors     
-nucleotide quality      
-length           

2. Evaluate Quality    

3. De novo transcriptome assembly (.fasta)   
3. A Evaluate assembly (compare to closely related species or Core Eukaryotic Genes N-50, # contigs)      
3. B Annotation      

4. Map reads to ref. transcriptome (.sam files; sequence alignment files)     

5. A Extract read count info (no of reads that map to each contig for each sample)      
5. B Identify SNP's      

6. A Differential gene expression analyses and co expression network analysis      
6. B Population genomics, genetic differentiation, pop. structure, demographic history, signatures of selection    

## Dunning et al. 2014 Paper discussion    

Unigene-alternative splice variant contigs are collapsed into "unigenes" (typically the longest)        
Picture in notebook   

Gene Ontology- Functional categories a gene is in (e.g. immune response gene)        

3. B Annotation   
Blast X to  
-nr (gene annotation)     
-uniprot database (curated protein database)    
-gene ontology   

Fisher's exact test (chi squared test)    
e.g. if 60K unigenes    
~10% of them are involved in cuticular hydrocarbon    

2k differentially expressed genes    
~20% of them are invovled in cuticular hydrocarbon (higher than predicted)     

Q plot -> y-axis ancestry, x-axis individuals of species. colors represent portion of where genome of individuals come from.       

K-number of groups- cross validation support 3 groups best (lowest cross validation error)   

Table2    
CS- increased expression in cold shock      
CS- increased expression in control     
DRR- Number of differentially expressed genes      
 
## Coding

De novo assembly on Trinity    

Go on server through PuTTY    

Cleaning on Trimmomatic PE   

We will evaluate, clean and evaluate some fastq files each.


Mine are 
09_5-14_S_2_R1.fq.gz
09_5-14_S_2_R2.fq.gz

09 is the individual number
5-14  is the date
S is sick (H would be a healthy individual)
2 (rating of sickness on a 0-5 scale, 0 being healthy, and 5 being dead)
"cp" command for copy ("cp XYZ ~/mydata/" means copy file XYZ into the mydata that is in home directory)

Use zcat to look at zipped file

zcat FILENAME | head

"head" gives us the first few lines (10?) instead of showing the whole file

Lines
open line with "J00160...." are the sequencing details (adaptors...etc)
sequence
letters giving the quality of the sequence above
quality score


Rest of info (about phred score...etc) on [Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-02-06_RNAseq_tutorial.html)     

------
<div id='id-section7'/>
### Entry 7: 02-08-2017    

## Info Update: Transcriptomics    

- Intro    
Why study wild systems?     
-Non model/non traditional model organisms      
-Silent genes responding to multiple stimuli      
-Novel transcripts w/o homologs in closely related model organisms     

- Methods    
Micro Array: easy for ecological analyses    
RNA-Seq: Genome-wide, more involved analyses     

- Questions   
-Questions about variation in gene expression          
-How gene expression affect phenotypes      
-Effect of environmental stimuli on gene expression      
Only 5 papers looked at all three (Venn Diagram in notebook about number of studies looking at each question)         

1. Gene expression variation at species, population, individual levels. Variation at different time or life history stages.      

Evolutionary Process    
 
-Gene expression heritable (naturally selected)       
-Epigenetics      
-Qst & Fst comparison     
-eQTL- expression quatitative trait loci mapping     
-Macroevolution: drift, bottleneck, selection    

2. Environmental stimuli on gene expression     
-abiotic stress    
-environmental heterogeneity    
-host-parasite interactions    
-selective biotic & abiotic interactions       
-flash freezing (to avoid stimuli) takes a snapshot of the moment     
-time course analysis- transcriptional response through time     

3. How gene expression affect phenotype     
-alternate phenotypes       
-manipulating transcription to make inferences about genotype-phenotype relationships -> transgenics (moving from correlation to causation): RNAi, CRISPR    

- Future      
-Combined micro arrays and RNA-seq (mostly used for model organisms)       
-Expand on databse for proposed ecological variations       

- Problems    
-Bias in signal     
-Heterologous arrays (relevant for micro arrays)      
-Polyploidy     
-RNA pooling    
-Statistical analyses       
-unannotated genes    

Great Question!!! *Is gene expression a phenotype itself or does it affect a phenotype?* 

## Coding:

Moving files to PC using WinSCP

Finish cleaning on Trimmomatic   
Start assembly

[Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-02-13_RNAseq_Mapping.html)

------
<div id='id-section8'/>
### Entry 8: 2017-02-13

## Info Update: RNA-seq 3

- Background
-Enables DE examination (inter-population individual)    
Disease Resistance
Mating Behavior
Adaptive significance

-Molecular Mechanisms -> Phenotypic/behavioral plasticity, migration patterns

-Limitations
Reference genome quality
Gene annotation availability
Expense per sample lib. prep.

- Issues
-Under utilization of biological replicates
Requiring independent library preparations
Doesn't include pooled samples
23/158 studies (15%) > 3 Bio replicates
Derive broad biological conclusions

-Prioritize seq depth over replication 

-Wide dynamic range of RNA-seq data noisy (Poisson counting error, technical variance, biological variance)

```
Glossary
- sequence coverage: the average # of reads that align/cover known reference bases
- read depth: total # bases sequenced/aligned at given reference base position
- power: probability of rejecting false null hypothesis
```

- General rules of thumb
1. Use more biological replicates instead of depth
2. Sequence depth >10 reads per transcript. ~10-20 million mapped reads/samples
3. 3 biological replicates per condition
4. Conduct a pilot experiment to know:
i) What is best/most powerful experiment that can be afforded
ii) what is the smallest fold change that can be detected


## Coding:

grep command and tutorial script used to know whether a read maps to a single place on the transcriptome or several. Important for gene expression studies since expression of a gene is based on the number of read counts. 

[Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-02-15_RNAseq_Map_Count.html)

Extracting read counts from SAM file

All SAM files put together using a bash script on the tutorial.

Cleaning up files and install Trinity and Bowtie

Instructions to start screen on PuTTY on Tutorial (can have things running in the background even when we are not on)

Install DESeq2 package in R. 

------
<div id='id-section9'/>
### Entry 9: 2017-02-15 and 2017-02-22

Class notes in physical notebook     
## Info Update on SNPs and Population genomics      

## Coding:

Using DESeq2 for gene expression analysis on R

[DESeq2 Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-2-22_RNAseqDGE.html)


------
<div id='id-section10'/>
### Entry 10: 2017-02-27
    

Class notes in physical notebook     

## Phylogenomicist Scott Edwards' visit    

------
<div id='id-section11'/>
### Entry 11: 2017-03-01

## Info Update: ME!! on heretic voodoo magic aka gene co-expression network analysis :)      

Notes in notebook (pictures posted soon)     

------
<div id='id-section12'/>
### Entry 12: 2017-03-06     

I was absent       

Refer to [Andrew's note](https://github.com/adnguyen/Notebooks_and_Protocols/blob/master/2017_Eco_Gen_ANBE_nb.md#id-section15) and [Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-06_Tutorials_PopGenomics1.html) for notes and population genomics coding       

------
<div id='id-section13'/>
### Entry 13: 2017-03-08      

Notes in physical notebook (pictures posted soon)      
## Info update on Rate of evolution       

## Coding:         

[Population Genomics Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-08_Tutorials_PopGenomics2.html)

vcf tools filtering strategy from last time       
```
$ vcftools --gzvcf SSW_byind.txt.vcf.gz --min-alleles 2 --max-alleles 2 --maf 0.02 --max-missing 0.8 --recode --out ~/SSW_all_biallelic.MAF0.02.Miss0.8    

$ gzip SSW_all_biallelic.MAF0.02.Miss0.8.recode.vcf      

$ vcftools --gzvcf SSW_all_biallelic.MAF0.02.Miss0.8.recode.vcf --hardy       
```

min-alleles 2 represents minimum number of alleles at each SNP set at 2     
max-alleles 2 represents the maximum number of alleles for each SNP set at 2(typically not common to get more than 2)     
maf 0.02 means that alleles with less than 2% frequency were ignored     
max-missing 0.8- include only sites that were present in 80% in samples     


Info on calculating diversity metrics in tutorial.     

Subset vcf file into healthy and sick     

Allele frequencies between H and S individuals and Fst calculated and plotted on R.     

Command line version of R can be used on server.     
```
R

R version 3.3.2 (2016-10-31) -- "Sincere Pumpkin Patch"

Copyright (C) 2016 The R Foundation for Statistical Computing

Platform: x86_64-redhat-linux-gnu (64-bit)

> fst <- read.table("HvS_OneSampPerInd.weir.fst",header=T)

> str(fst)

'data.frame':   442 obs. of  3 variables:

 $ CHROM : Factor w/ 111 levels "TRINITY_DN35598_c0_g1_TRINITY_DN35598_c0_g1_i1_g.5802_m.5802",..: 65 65 100 100 100 100 100 100 88 88 ...

 $ POS : int  4566 4665 978 1404 1722 3426 3729 3912 115 141 ...

 $ WEIR_AND_COCKERHAM_FST: num  0.0305 0.0085 0.0305 -0.0188 0.0732 ...
```

Saving it to pdf    

```
 

> pdf("HvS_Fst.pdf")

> hist(fst$WEIR_AND_COCKERHAM_FST, breaks=20, col="red")

> dev.off()

null device 
```


------
<div id='id-section14'/>
### Entry 14: 2017-03-20      

Notes on physical notebook       
## Info update on Local and Global Ancestry      

## Coding:    

Finishing up on diversity and allele frequency calculations      

[Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-20_Tutorials_PopGenomics3.html)        

Use grep to combine HS and SS into one group     

Use of vcf tools to refilter    

Bring data to our computer for analysis and plotting on R     
[R script](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-20_populationgenomics3_R.html)     


------
<div id='id-section15'/>
### Entry 15: 2017-03-22     

Absent     
Notes, Info update on inferrence divergence history and discussion available on [Andrew's github](https://github.com/adnguyen/Notebooks_and_Protocols/blob/master/2017_Eco_Gen_ANBE_nb.md#id-section18)     

## Coding      

[R script](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-22_pop4_followalong.html)     


------
<div id='id-section16'/>
### Entry 16:2017-03-27      

I was absent     
Notes, Info update on selective sweep, discussion on [Andrew's Github](https://github.com/adnguyen/Notebooks_and_Protocols/blob/master/2017_Eco_Gen_ANBE_nb.md#id-section19)
## Coding     
[Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-22_Tutorials_PopulGenomics4.html)


------
<div id='id-section17'/>
### Entry 17: 2017-03-29     

Notes in notebook     
## Info update on detecting local adaptation      

## Coding     
Admixture      
[Tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-22_Tutorials_PopulGenomics4.html) and [R script](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-03-22_pop4_followalong.html)    


------
<div id='id-section18'/>
### Entry 18: 2017-04-03    

Notes in Notebook 9to be posted soon)    
## Karl Fetter on Fst    
[Notes on Fst](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-04-03_Popgenomics5_outlier_annotations_DAPC_OutFLANK.html) and [OUTFLANK](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-04-03_OUTFLANK-yourself.html)

------
<div id='id-section19'/>
### Entry 19: 2017-04-05    

Note in Notebook     

## Info update on Annotation      

[Tutorial and info](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-04-05_annotation_tutorial3.html)    


------
<div id='id-section20'/>
### Entry 20:2017-04-08     

Notes in notebook     

## Info update on Microbiome       

## Coding     
Microbiome with Melanie
[Analyzing 16S amplicon tutorial](https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-04-10_16sAmipliconSeqData.html)     

------
<div id='id-section21'/>
### Entry 21:
------
<div id='id-section22'/>
### Entry 22:
------
<div id='id-section23'/>
### Entry 23:
------
<div id='id-section24'/>
### Entry 24:
------
<div id='id-section25'/>
### Entry 25:
------
<div id='id-section26'/>
### Entry 26:
------
<div id='id-section27'/>
### Entry 27:
------
<div id='id-section28'/>
### Entry 28:
------
<div id='id-section29'/>
### Entry 29: Code for HW2 

## Code for intertidal

```
data=which(colData[,5]=="int")
A=colData[data,]
B=(countData[,1:48])
dim(colData)
dim(A)
dim(colData)
dim(B)

#model code
dds <- DESeqDataSetFromMatrix(countData = B, colData = A, design = ~ + health)
dim(dds)
dds <- dds[ rowSums(counts(dds)) > 100, ]
dim(dds)
dds <- dds[sample(nrow(dds)), ]
dim(dds)
colData(dds)$health <- factor(colData(dds)$health, levels=c("H","S")) #sets that "healthy is the reference
dds <- DESeq(dds) 
res <- results(dds)
res <- res[order(res$padj),]
head(res)


summary(res)
plotMA(res, main="DESeq2", ylim=c(-3,3))


## Check out one of the genes to see if it's behaving as expected....I put the most significantly differentially expressed gene in these parts
d <- plotCounts(dds, gene="TRINITY_DN43080_c1_g1_TRINITY_DN43080_c1_g1_i3_g.14110_m.14110", intgroup=(c("health","day","location")), returnData=TRUE)
d
p <- ggplot(d, aes(x= health, y=count, shape = location)) + theme_minimal() + theme(text = element_text(size=20), panel.grid.major = element_line(colour = "grey"))
p <- p + geom_point(position=position_jitter(w=0.3,h=0), size = 3) + ylim(0,3000)
p


## Check out one of the genes to see interaction between score, health and expression....
d <- plotCounts(dds, gene="TRINITY_DN43080_c1_g1_TRINITY_DN43080_c1_g1_i3_g.14110_m.14110", intgroup=(c("health","score","location")), returnData=TRUE)
d
p <- ggplot(d, aes(x= score, y=count, shape = health, color = location)) + theme_minimal() + theme(text = element_text(size=20), panel.grid.major = element_line(colour = "grey"))
p <- p + geom_point(position=position_jitter(w=0.3,h=0), size = 3) 
p

p <- ggplot(d, aes(x=score, y=count, color=health, group=health)) 
p <- p +  geom_point() + stat_smooth(se=FALSE,method="loess") +  scale_y_log10()
p
```

## Code for Subtidal   

```
data=which(colData[,5]=="sub")
C=colData[data,]
D=(countData[,49:77])
dim(colData)
dim(C)
dim(colData)
dim(D)

#model code
dds <- DESeqDataSetFromMatrix(countData = D, colData = C, design = ~ + health)
dim(dds)
dds <- dds[ rowSums(counts(dds)) > 100, ]
dim(dds)
dds <- dds[sample(nrow(dds)), ]
dim(dds)
colData(dds)$health <- factor(colData(dds)$health, levels=c("H","S")) #sets that "healthy is the reference
dds <- DESeq(dds) 
res <- results(dds)
res <- res[order(res$padj),]
head(res)

summary(res)
plotMA(res, main="DESeq2", ylim=c(-7,7))

## Check out one of the genes to see if it's behaving as expected....I put the most significantly differentially expressed gene in these parts
d <- plotCounts(dds, gene="TRINITY_DN42073_c0_g1_TRINITY_DN42073_c0_g1_i1_g.12173_m.12173", intgroup=(c("health","day","location")), returnData=TRUE)
d
p <- ggplot(d, aes(x= health, y=count, shape = location)) + theme_minimal() + theme(text = element_text(size=20), panel.grid.major = element_line(colour = "grey"))
p <- p + geom_point(position=position_jitter(w=0.3,h=0), size = 3) + ylim(0,200)
p


## Check out one of the genes to see interaction between score, health and expression....
d <- plotCounts(dds, gene="TRINITY_DN42073_c0_g1_TRINITY_DN42073_c0_g1_i1_g.12173_m.12173", intgroup=(c("health","score","location")), returnData=TRUE)
d
p <- ggplot(d, aes(x= score, y=count, shape = health, color = location)) + theme_minimal() + theme(text = element_text(size=20), panel.grid.major = element_line(colour = "grey"))
p <- p + geom_point(position=position_jitter(w=0.3,h=0), size = 3) 
p

p <- ggplot(d, aes(x=score, y=count, color=health, group=health)) 
p <- p +  geom_point() + stat_smooth(se=FALSE,method="loess") +  scale_y_log10()
p

#Code for both intertidal and subtidal (w/out interaction)

E=(colData) 
dim(E)
G=(countData)
dim(G)

#model code
dds <- DESeqDataSetFromMatrix(countData = G, colData = E, design = ~ location + health)
dim(dds)
dds <- dds[ rowSums(counts(dds)) > 100, ]
dim(dds)
dds <- dds[sample(nrow(dds)), ]
dim(dds)
colData(dds)$health <- factor(colData(dds)$health, levels=c("H","S")) #sets that "healthy is the reference
dds <- DESeq(dds) 
res <- results(dds)
res <- res[order(res$padj),]
sigres <-res[res$padj>0.05,]
head(res)

summary(res)

plotMA(res, main="DESeq2", ylim=c(-4,4))

## Check out one of the genes to see if it's behaving as expected....I put the most significantly differentially expressed gene in these parts
d <- plotCounts(dds, gene="TRINITY_DN43080_c1_g1_TRINITY_DN43080_c1_g1_i3_g.14110_m.14110", intgroup=(c("health","day","location")), returnData=TRUE)
d
p <- ggplot(d, aes(x= health, y=count, shape = location)) + theme_minimal() + theme(text = element_text(size=20), panel.grid.major = element_line(colour = "grey"))
p <- p + geom_point(position=position_jitter(w=0.3,h=0), size = 3) + ylim(0,2000)
p


## Check out one of the genes to see interaction between score, health and expression....
d <- plotCounts(dds, gene="TRINITY_DN43080_c1_g1_TRINITY_DN43080_c1_g1_i3_g.14110_m.14110", intgroup=(c("health","score","location")), returnData=TRUE)
d
p <- ggplot(d, aes(x= score, y=count, shape = health, color = location)) + theme_minimal() + theme(text = element_text(size=20), panel.grid.major = element_line(colour = "grey"))
p <- p + geom_point(position=position_jitter(w=0.3,h=0), size = 3) 
p

p <- ggplot(d, aes(x=score, y=count, color=health, group=health)) 
p <- p +  geom_point() + stat_smooth(se=FALSE,method="loess") +  scale_y_log10()
p
```
------
<div id='id-section30'/>
### Entry 30: Code for HW3   

## Code for minor allele frequency change (MAF=0.04)   
```
# Set your working directory to where you downloaded your results files:
#setwd("C:/Users/Muhammad/Desktop/Ecological Genomics/HW3/Minor Allele frequency changed")

list.files() # Do you see your downloaded files there? If not, double check to make sure you've set your working directory to the right spot

# We'll need to install 2 packages to work with the SNP data:
#install.packages("vcfR") # reads in vcf files and proides tools for file conversion 
#install.packages("adegenet") # pop-genetics package with some handy routines, including PCA and other multivariate methods (DAPC)

# ...and load the libraries
library(adegenet)
library(vcfR)


vcf1<-read.vcfR("SSW_all_biallelic.MAF0.04.Miss0.8.recode.vcf")
#vcf1 <- read.vcfR("SSW_all_biallelic.MAF0.04.Miss0.8.recode.vcf")

# The adegenet package uses a highly efficient way of storing large SNP datasets in R called a "genlight" object. The following function creates a genlight object from your vcf:
gl1 <- vcfR2genlight(vcf1)
print(gl1) # Looks good! Right # of SNPs and individuals!

# For info, try:
gl1$ind.names
gl1$loc.names[1:10]

# Notice there's nothing in the field that says "pop"? Let's fix that...
ssw_meta <- read.table("ssw_healthloc.txt", header=T) # read in the metadata
ssw_meta <- ssw_meta[order(ssw_meta$Individual),] # sort it by Individual ID

# Confirm the ID's are ordered the same in gl1 and ssw_meta:
gl1$ind.names
ssw_meta$Individual

gl1$pop <- ssw_meta$Location # assign locality info
gl1$other <- as.list(ssw_meta$Trajectory) # assign disease status


# WE can explore the structure of our SNP data using the glPlot function, which gives us a sample x SNP view of the VCF file
glPlot(gl1, posi="bottomleft")

# Now, let's compute the PCA on the SNP genotypes and plot it:
pca1 <- glPca(gl1, nf=4, parallel = F) # nf = number of PC axes to retain (here, 4)
pca1 # prints summary

# Plot the individuals in SNP-PCA space, with locality labels:
plot(pca1$scores[,1], pca1$scores[,2], 
     cex=2, pch=20, col=gl1$pop, 
     xlab="Principal Component 1", 
     ylab="Principal Component 2", 
     main="PCA on SSW data (Freq missing=20%; 2493 SNPs)")
legend("topleft", 
       legend=unique(gl1$pop), 
       pch=20, 
       col=c("black", "red"))

# Perhaps we want to show disease status instead of locality:
plot(pca1$scores[,1], pca1$scores[,2], 
     cex=2, pch=20, col=as.factor(unlist(gl1$other)), 
     xlab="Principal Component 1", 
     ylab="Principal Component 2", 
     main="PCA on SSW data (Freq missing=20%; 2493 SNPs)")
legend("topleft", 
       legend=unique(as.factor(unlist(gl1$other))), 
       pch=20, 
       col=as.factor(unique(unlist(gl1$other))))

# Which SNPs load most strongly on the 1st PC axis?
loadingplot(abs(pca1$loadings[,1]),
            threshold=quantile(abs(pca1$loadings), 0.999))
# Get their locus names
gl1$loc.names[which(quantile(abs(pca1$loadings))>0.999)]

threshold<-quantile(abs(pca1$loadings),0.999)

gl1$loc.names[which(abs(pca1$loadings)>threshold)]

gl1$loc.names[which(quantile(abs(pca1$loadings),0.999)>0.0770)]
```

## Code for change in max missing (missing 1.0)
```
setwd("C:/Users/Muhammad/Desktop/Ecological Genomics/HW3/Site missing")

list.files() # Do you see your downloaded files there? If not, double check to make sure you've set your working directory to the right spot

# We'll need to install 2 packages to work with the SNP data:
#install.packages("vcfR") # reads in vcf files and proides tools for file conversion 
#install.packages("adegenet") # pop-genetics package with some handy routines, including PCA and other multivariate methods (DAPC)

# ...and load the libraries
library(adegenet)
library(vcfR)



vcf1<-read.vcfR("SSW_all_biallelic.MAF0.02.Miss1.0.recode.vcf")
#vcf1 <- read.vcfR("SSW_all_biallelic.MAF0.02.Miss1.0.recode.vcf")

# The adegenet package uses a highly efficient way of storing large SNP datasets in R called a "genlight" object. The following function creates a genlight object from your vcf:
gl1 <- vcfR2genlight(vcf1)
print(gl1) # Looks good! Right # of SNPs and individuals!

# For info, try:
gl1$ind.names
gl1$loc.names[1:10]

# Notice there's nothing in the field that says "pop"? Let's fix that...
ssw_meta <- read.table("ssw_healthloc.txt", header=T) # read in the metadata
ssw_meta <- ssw_meta[order(ssw_meta$Individual),] # sort it by Individual ID

# Confirm the ID's are ordered the same in gl1 and ssw_meta:
gl1$ind.names
ssw_meta$Individual

gl1$pop <- ssw_meta$Location # assign locality info
gl1$other <- as.list(ssw_meta$Trajectory) # assign disease status


# WE can explore the structure of our SNP data using the glPlot function, which gives us a sample x SNP view of the VCF file
glPlot(gl1, posi="bottomleft")

# Now, let's compute the PCA on the SNP genotypes and plot it:
pca1 <- glPca(gl1, nf=4, parallel = F) # nf = number of PC axes to retain (here, 4)
pca1 # prints summary

# Plot the individuals in SNP-PCA space, with locality labels:
plot(pca1$scores[,1], pca1$scores[,2], 
     cex=2, pch=20, col=gl1$pop, 
     xlab="Principal Component 1", 
     ylab="Principal Component 2", 
     main="PCA on SSW data (Freq missing=0%; 1494 SNPs)")
legend("topleft", 
       legend=unique(gl1$pop), 
       pch=20, 
       col=c("black", "red"))

# Perhaps we want to show disease status instead of locality:
plot(pca1$scores[,1], pca1$scores[,2], 
     cex=2, pch=20, col=as.factor(unlist(gl1$other)), 
     xlab="Principal Component 1", 
     ylab="Principal Component 2", 
     main="PCA on SSW data (Freq missing=0%; 1494 SNPs)")
legend("topleft", 
       legend=unique(as.factor(unlist(gl1$other))), 
       pch=20, 
       col=as.factor(unique(unlist(gl1$other))))

# Which SNPs load most strongly on the 1st PC axis?
loadingplot(abs(pca1$loadings[,1]),
            threshold=quantile(abs(pca1$loadings), 0.999))
# Get their locus names
gl1$loc.names[which(quantile(abs(pca1$loadings))>0.999)]

threshold<-quantile(abs(pca1$loadings),0.999)

gl1$loc.names[which(abs(pca1$loadings)>threshold)]

gl1$loc.names[which(quantile(abs(pca1$loadings),0.999)>0.0770)]
```
------
<div id='id-section31'/>
### Entry 31: Scripts for Project (Microbiome)

```     
library("phyloseq"); packageVersion("phyloseq")    
library("DESeq2")
packageVersion("DESeq2")
library("ggplot2")
theme_set(theme_bw())

#Import the OTU table
otutable <- import_biom(BIOMfilename = 'otu_table_mc2_w_tax_no_pynast_failures_no_chimeras_frequency_filtered.biom', 
                        treefilename = 'rep_set_no_chimeras.tre', 
                        parseFunction = parse_taxonomy_greengenes)

#The warnings are ok. There is 1 warning for every OTU that doesn't have a taxonomy assignment

#Import the mapping file
mapping <- import_qiime_sample_data(mapfilename = 'R_map.txt')

#Merge the mapping file to the OTU table
phylo <- merge_phyloseq(otutable, mapping)

#Check to make sure the imports worked
phylo

###############################################################################
####Testing for differentially expressed OTUs between Sick and Healthy samples
####while controlling for the repeated measures on each individual
###############################################################################

##For now we're only going to look at differences between Sick and Healthy samples and 
##remove the Dead samples from the analysis
phylo_subset = subset_samples(phylo, Phenotype != "Dead")

##We want to numbers of each individual to be a factor. Right now it's an integer so we have to change that.
class(sample_data(phylo_subset)$individual)
sample_data(phylo_subset)$individual<-factor(sample_data(phylo_subset)$individual)

##Phyloseq's wrapper to get OTU data into DESeq
pheno <- phyloseq_to_deseq2(phylo_subset, ~ individual + Phenotype)

##Run DESeq. This command takes a bit of time
pheno_deseq_test <- DESeq(pheno, test="Wald")

##Get results from DESeq
pheno_results <- results(pheno_deseq_test)

##Have a look at the summary
summary(pheno_results)
head(pheno_results)

##Make table with OTUs padj<0.05
alpha <- 0.05
pheno_sigtab <- pheno_results[which(pheno_results$padj < alpha), ]

##Add taxa info to that table
pheno_sigtab <- cbind(as(pheno_sigtab, "data.frame"), as(tax_table(phylo)[rownames(pheno_sigtab), ], "matrix"))
tail(pheno_sigtab)

##Save the table to your desktop
write.table(pheno_sigtab, "DE_OTU_sickk_vs_healthy.txt", sep="\t")

##plot taxa of all otus padj<0.05 using ggplots2
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}

x <- tapply(pheno_sigtab$log2FoldChange, pheno_sigtab$Phylum, function(x) max(x))
x <- sort(x, TRUE)
pheno_sigtab$Phylum = factor(as.character(pheno_sigtab$Phylum), levels=names(x))
# Family order
x <- tapply(pheno_sigtab$log2FoldChange, pheno_sigtab$Family, function(x) max(x))
x <- sort(x, TRUE)
pheno_sigtab$Family = factor(as.character(pheno_sigtab$Family), levels=names(x))
ggplot(pheno_sigtab, aes(x=Family, y=log2FoldChange, color=Phylum)) + geom_point(size=3) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

##plot counts across days for individual otus
title <- "OTU Number"
data <- plotCounts(pheno_deseq_test, "OTU Number" , 
                   intgroup=c("Day","Phenotype"), returnData=TRUE)
ggplot(data, aes(x=Day, y=count, color=Phenotype, group=Phenotype)) + 
  geom_point() + stat_smooth(se=FALSE,method="loess") +  scale_y_log10() + ggtitle(title)     
```
------
<div id='id-section32'/>
### Entry 32:
------
<div id='id-section33'/>
### Entry 33:
------
<div id='id-section34'/>
### Entry 34:
------
<div id='id-section35'/>
### Entry 35:
------
<div id='id-section36'/>
### Entry 36:
------
<div id='id-section37'/>
### Entry 37:
------
<div id='id-section38'/>
### Entry 38:
------
<div id='id-section39'/>
### Entry 39:
------
<div id='id-section40'/>
### Entry 40:
------
<div id='id-section41'/>
### Entry 41:
------
<div id='id-section42'/>
### Entry 42:
------
<div id='id-section43'/>
### Entry 43:
------
<div id='id-section44'/>
### Entry 44:
------
<div id='id-section45'/>
### Entry 45:
------
<div id='id-section46'/>
### Entry 46:
------
<div id='id-section47'/>
### Entry 47:
------
<div id='id-section48'/>
### Entry 48:
------
<div id='id-section49'/>
### Entry 49:
------
<div id='id-section50'/>
### Entry 50:
------
<div id='id-section51'/>
### Entry 51:
------
<div id='id-section52'/>
### Entry 52:
------
<div id='id-section53'/>
### Entry 53:
------
<div id='id-section54'/>
### Entry 54:
------
<div id='id-section55'/>
### Entry 55:
------
<div id='id-section56'/>
### Entry 56:
------
<div id='id-section57'/>
### Entry 57:
------
<div id='id-section58'/>
### Entry 58:
------
<div id='id-section59'/>
### Entry 59:
------
<div id='id-section60'/>
### Entry 60:

------
